### PR TITLE
Database templates

### PR DESCRIPTION
**Pull request by leynos**
## Summary by Sourcery

Add database lifecycle and template support for fast, isolated PostgreSQL testing, including shared cluster fixtures, temporary databases, and supporting documentation and utilities.

New Features:
- Expose APIs on TestCluster and TestClusterConnection to create, clone-from-template, drop, and probe PostgreSQL databases on a running test cluster.
- Introduce a TemporaryDatabase RAII guard to automatically clean up per-test databases, with optional forced drop that terminates active connections.
- Provide a process-global shared test cluster accessor and rstest fixture to reuse a single PostgreSQL cluster across tests for faster execution.
- Add a directory hashing utility to derive content-based template names from migration trees.

Enhancements:
- Document shared-cluster, template-database, and cleanup patterns in the user guide, README, and a new execution plan, including performance guidance and roadmap conventions.
- Refine documentation standards with roadmap and ADR formatting guidance and fix a Mermaid diagram type in the documentation style guide.
- Export new testing and cluster helpers through the public API surface for easier consumption by downstream crates.

Documentation:
- Add extensive user-guide coverage of database lifecycle management, template databases, shared cluster usage, performance trade-offs, and cleanup strategies.
- Document ADR 002 describing the design and rationale for template database support and fast test isolation, plus an execution plan for implementation.
- Update the README with a high-level overview and example of using shared clusters and template databases for fast tests.
- Extend the documentation style guide with roadmap and ADR conventions and correct the Mermaid flowchart syntax.

Tests:
- Add behavioural tests and feature files covering database lifecycle operations and TestCluster delegation to ensure correct creation, deletion, and error handling.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/database_lifecycle.rs](https://github.com/leynos/pg-embedded-setup-unpriv/pull/43/files/c7e4b40319cf57f3aa9329b86642b0988c3871c8#diff-f0efa61d8db06b373dab7e5178274ff5cac449f6ef74bfc06849ec4c2ddb87b1)

Comment on lines +161 to +175

~~~rust
fn when_database_created(world: &DatabaseWorldFixture) -> Result<()> {
    let world_cell = borrow_world(world)?;
    if world_cell.borrow().is_skipped() {
        return Ok(());
    }
    let result = world_cell
        .borrow()
        .cluster()?
        .connection()
        .create_database(TEST_DB_NAME);
    if let Err(err) = result {
        world_cell.borrow_mut().create_error = Some(err.to_string());
    }
    Ok(())
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72457/delta?repo-id=1225432&review-id=43&biomarker=Code+Duplication&filename=tests%2Fdatabase_lifecycle.rs&method=)
The module contains 12 functions with similar structure: then_database_does_not_exist,then_database_does_not_exist_via_delegation,then_database_exists,then_database_exists_via_delegation and 8 more functions

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

⚠️ Partially Resolved (1 issue)

    Comment 11 (unit/behavioural tests): Behavioural tests exist in tests/database_lifecycle.rs with 4 scenarios, but they don't cover temporary_database, temporary_database_from_template, or ensure_template_exists (see unresolved Comments 4, 6, and 12).

❌ Not Resolved (6 issues)

    Overall Comment 2 (macro for delegation): The 7 delegation methods on TestCluster are still individually implemented rather than using a macro to reduce duplication.

    Overall Comment 3 (admin_client reuse): temporary_database.rs still calls Client::connect(&self.admin_url, NoTls) directly in two places (lines 131, 165) instead of reusing the admin_client helper.

    Comment 4 (temporary database test scenarios): No behavioural tests exist for temporary_database or temporary_database_from_template RAII semantics.

    Comment 6 (ensure_template_exists scenarios): No scenarios test ensure_template_exists or its per-template locking behaviour.

    Comment 7 (hash directory tests): Only 3 basic tests exist. Missing tests for nested directories and empty directories.

    Comment 8 (concurrent-safe terminology): README.md line 129 still uses "concurrent-safe" instead of "concurrency-safe".

    Comment 12 (TemporaryDatabase drop tests): No unit or behavioural tests validate the drop_database, force_drop, or automatic Drop behaviour.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

⚠️ Partially Resolved (1 issue)

    Comment 12 (TemporaryDatabase drop tests): Behavioural tests exist for the automatic Drop cleanup (scenario 4), but unit tests are still missing for:
        force_drop() method (terminates active connections before dropping)
        Explicit drop_database(self) method
        Drop trait failure handling (logging on cleanup failure)

    Only accessor tests (temporary_database_accessors) exist in the unit test module.

File: docs/users-guide.md [hunk 1]
```diff
@@ -207,6 +258,240 @@ assert!(url.starts_with("postgresql://"));
...
+
+### Performance comparison
+
+The following table compares test isolation approaches:
+
+| Approach                        | Bootstrap | Per-test overhead | Isolation |
+| ------------------------------- | --------- | ----------------- | --------- |
+| Per-test `TestCluster`          | Per test  | 20-30 seconds     | Full      |
+| Shared cluster, fresh database  | Once      | 1-5 seconds       | Database  |
+| Shared cluster, template clone  | Once      | 10-50 ms          | Database  |
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API; otherwise, please provide an agent prompt for the remaining work.

File: tests/features/database_lifecycle.feature [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    Then the template database exists
+    And the setup function was called exactly once
+    When ensure_template_exists is called again for the same template
+    Then the setup function was still called exactly once
+
+  Scenario: Create database from template clones template
+    Given a sandboxed TestCluster is running
+    When a template database is created and populated
+    And a database is created from the template
+    Then the cloned database exists
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/features/database_lifecycle.feature]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API; otherwise, please provide an agent prompt for the remaining work.

File: src/test_support/hash.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            // Hash file contents
+            let mut file = fs::File::open(&path).map_err(|e| {
+                color_eyre::eyre::eyre!("failed to open file '{}': {e}", path.display())
+            })?;
+            let mut buffer = Vec::new();
+            file.read_to_end(&mut buffer).map_err(|e| {
+                color_eyre::eyre::eyre!("failed to read file '{}': {e}", path.display())
+            })?;
+            hasher.update(&buffer);
+        }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/test_support/hash.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API; otherwise, please provide an agent prompt for the remaining work.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/cluster/lifecycle.rs](https://github.com/leynos/pg-embedded-setup-unpriv/pull/43/files/7287bd0d9da88796a8f6690463625210bab8f9ee#diff-1fb5f4309db27b30aecd7c45c5f291917eaeec3e86ad88bc41479f5437994756)

Comment on lines +48 to +57

~~~rust
    pub fn create_database(&self, name: &str) -> BootstrapResult<()> {
        let _span = info_span!("create_database", db = %name).entered();
        let mut client = self.admin_client()?;
        let escaped = escape_identifier(name);
        let sql = format!("CREATE DATABASE \"{escaped}\"");
        client
            .batch_execute(&sql)
            .wrap_err(format!("failed to create database '{name}'"))
            .map_err(crate::error::BootstrapError::from)
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72457/delta?repo-id=1225432&review-id=43&biomarker=Code+Duplication&filename=src%2Fcluster%2Flifecycle.rs&method=)
The module contains 2 functions with similar structure: TestClusterConnection.create_database,TestClusterConnection.drop_database

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/cluster/lifecycle.rs](https://github.com/leynos/pg-embedded-setup-unpriv/pull/43/files/7287bd0d9da88796a8f6690463625210bab8f9ee#diff-1fb5f4309db27b30aecd7c45c5f291917eaeec3e86ad88bc41479f5437994756)

Comment on file

~~~rust
//! Database lifecycle operations for `TestClusterConnection`.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/72457/delta?repo-id=1225432&review-id=43&biomarker=String+Heavy+Function+Arguments&filename=src%2Fcluster%2Flifecycle.rs&method=)
In this module, 52.9% of all arguments to its 8 functions are strings. The threshold for string arguments is 39.0%

**Comment by leynos** [comment]
@coderabbitai review

File: src/cluster/worker_operation.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+use crate::TestBootstrapSettings;
+
+#[doc(hidden)]
+/// Identifies worker lifecycle operations executed via the helper binary.
+#[derive(Clone, Copy)]
+pub enum WorkerOperation {
+    Setup,
+    Start,
+    Stop,
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/cluster/worker_operation.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API; otherwise, please provide an agent prompt for the remaining work.
